### DISCUSSION TITLE
OpenGlControlBase and GL.DrawArrays(GL_LINE_STRIP, ...)

### DISCUSSION THREAD

**Discussion by kkmserver** [AvaloniaUI/Avalonia#15787] [Q&A] [answered]
We are writing an application that should work under Linux and Windows.
The application must display the route of the tool.

GL.DrawArrays(GL_LINE_STRIP, ...) is ideal for these purposes - minimum data, maximum performance.

But for some reason OpenGlControlBase does not have GL_LINE_STRIP. GL_TRIANGLES only.
We ourselves added:
'public const int GL_LINE_STRIP = 3;'

And on Windows GL.DrawArrays(GL_LINE_STRIP,..) worked perfectly.
But in Linux it shows a blank screen. :(
There are no errors in gl.GetError() (except for the problem described below)

Questions:
1. What determines the ability of GL.DrawArrays(GL_LINE_STRIP,..) to work?
2. Is there any way to make this work on Linux?
3. Or can this be solved some other way?

Also: on Linux in the official example: ControlCatalogStandalone 
in: OpenGlPage.axaml.cs
in function: 'protected override unsafe void OnOpenGlInit(GlInterface GL)' 
on the line (in the very first OpenGL API call): CheckError(GL);
immediately gives error 1280.
But the teapot is rendered.
I have the same thing on Linux.

Linux - stock distribution of Ubuntu 24 - which runs on an Intel Celeron J6712
AvaloniaUI 11.0.10

**Comment by kekekeks** [comment, marked as answer]
Our `GlInterface` class wraps the bare minimum required for avalonia and some samples to work. It was never supposed to be a full OpenGL binding. 

You need to use proper bindings like Silk.NET or define required GL constants and call GetProcAddress manually.

1) According to https://docs.gl/gl3/glDrawArrays / https://docs.gl/es2/glDrawArrays line strip should work with both ANGLE-powered GLES2 and native Linux OpenGL 3.2+,  but you might need some extra setup with gl state (like intermediate AA texture).
2) ^
3) You can try using skia api lease feature and call SKCanvas.DrawPoints

**Reply by kkmserver** [reply]
Thanks, I'll try to figure it out...

**Reply by FSwom** [reply]
Is there an example anywhere how "define required GL constants and call GetProcAddress manually" should look like?

**Reply by Coloryr** [reply]
Use silk.opengl
```
var silkgl = GL.GetApi(gl.GetProcAddress)
```
or
```
public delegate void Func1(int a, int b);
public Func1 GLBlendFunc = Marshal.GetDelegateForFunctionPointer<Func1>(gl.GetProcAddress("glBlendFunc"));
```
get you want.
